include/boost/geometry/formulas/karney_direct.hpp

```diff
@@ -0,0 +1,290 @@
+
+        Azi azi12 = azimuth12;
+        math::normalize_angle<degree, Azi>(azi12);
+
+        Dist const dist_c0 = 0;
+
+        if (math::equals(distance, dist_c0) || distance < dist_c0)
+        {
+            result.lon2 = lon1;
+            result.lat2 = lat1;
+            return result;
+        }
+
+        CT const c0 = 0;
+        CT const c1 = 1;
+        CT const c2 = 2;
+
+        CT const b = CT(get_radius<2>(spheroid));
+        CT const f = formula::flattening<CT>(spheroid);
+        CT const one_minus_f = c1 - f;
+        CT const two_minus_f = c2 - f;
+
+        CT const n = f / two_minus_f;
+        CT const e2 = f * two_minus_f;
+        CT const ep2 = e2 / math::sqr(one_minus_f);
+
+        CT sin_alpha1, cos_alpha1;
+        math::sin_cos_degrees<CT>(math::round_angle<CT>(azi12), sin_alpha1, cos_alpha1);
+
+        // Find the reduced latitude.
+        CT sin_beta1, cos_beta1;
+        math::sin_cos_degrees<CT>(math::round_angle<CT>(lat1), sin_beta1, cos_beta1);
+        sin_beta1 *= one_minus_f;
+
+        math::normalize_values<CT>(sin_beta1, cos_beta1);
+
+        cos_beta1 = std::max(sqrt(std::numeric_limits<CT>::min()), cos_beta1);
+
+        // Obtain alpha 0 by solving the spherical triangle.
+        CT const sin_alpha0
+            = sin_alpha1 * cos_beta1;
+        CT const cos_alpha0
+            = boost::math::hypot(cos_alpha1, sin_alpha1 * sin_beta1);
+
+        CT const k2 = math::sqr(cos_alpha0) * ep2;
+
+        CT const epsilon = k2 / (c2 * (c1 + sqrt(c1 + k2)) + k2);
+
+        // Find the coefficients for A1 by computing the
+        // series expansion using Horner scehme.
+        CT const expansion_A1
+            = series_expansion::evaluate_series_A1<CT, SeriesOrder>(epsilon);
+
+        // Index zero element of coeffs_C1 is unused.
+        CT coeffs_C1[SeriesOrder + 1];
+        series_expansion::evaluate_coeffs_C1<CT, SeriesOrder>(epsilon, coeffs_C1);
+
+        // Tau is an integration variable.
+        CT const tau12 = distance / (b * (c1 + expansion_A1));
+
+        CT const sin_tau12 = sin(tau12);
+        CT const cos_tau12 = cos(tau12);
+
+        CT sin_sigma1 = sin_beta1;
+        CT sin_omega1 = sin_alpha0 * sin_beta1;
+
+        CT cos_sigma1, cos_omega1;
+        cos_sigma1 = cos_omega1 = sin_beta1 != c0 || cos_alpha1 != c0 ? cos_beta1 * cos_alpha1 : c1;
+        math::normalize_values<CT>(sin_sigma1, cos_sigma1);
+
+        CT const B11 =
+            series_expansion::sin_cos_series<CT, SeriesOrder>(sin_sigma1, cos_sigma1, coeffs_C1);
+        CT const sin_B11 = sin(B11);
+        CT const cos_B11 = cos(B11);
+
+        CT const sin_tau1
+            = sin_sigma1 * cos_B11 + cos_sigma1 * sin_B11;
+        CT const cos_tau1
+            = cos_sigma1 * cos_B11 - sin_sigma1 * sin_B11;
+
+        // Index zero element of coeffs_C1p is unused.
+        CT coeffs_C1p[SeriesOrder + 1];
+        series_expansion::evaluate_coeffs_C1p<CT, SeriesOrder>(epsilon, coeffs_C1p);
```

Comment:
@awulkiew Yes, epsilon is only a parameter.

Reiterating what you said above, in case I use `boost::array`, I won't have to explicitly provide the template parameters, but they will still be part of the function header, right?

This works fine for most cases, however, in some cases the array size is defined as below:
```cpp
size_t const coeffs_C3_size = (SeriesOrder * (SeriesOrder - 1)) / 2;
```
The array can be defined as:
```cpp
boost::array<CT, coeffs_C3_size> coeffs_C3x;
```
Which is then passed to this function:
```cpp
template <typename CT, size_t SeriesOrder>
static inline void evaluate_coeffs_C3x(CT const& n, boost::array<CT, SeriesOrder>& c)
{ ... }
```
In this case, the `SeriesOrder` template parameter will be the updated array size, hence the `switch` statement will not have any effect, since the value is out of bound.

One solution is to pass an additional argument to the function, which represents `SeriesOrder`. Another way to solve this is by converting the series expansions functions to class methods, and storing `SeriesOrder` as a class member. Please let me know if a there's a better alternative.

I have [this commit](https://github.com/BoostGSoC18/geometry/commit/1972bcda3e5c204f2a0d0bda66239725828f1306), which might make it easier to follow what I am trying to say above.

Reply:
@adl1995 Yes the most simple solution is to pass `SeriesOrder` as separate parameter:

    template <size_T SeriesOrder, typename CT, size_t N>
    inline void evaluate_coeffs_C3x(boost::array<CT, N>& c, CT const& n)
    {
        BOOST_MPL_ASSERT(N == (SeriesOrder * (SeriesOrder - 1)) / 2);
        // ...
    }

but then the caller of this function has to know what's the required size of array for passed `SeriesOrder`. It would be better to have it all in one place. Would it have sense to implement several representations of coefficient containers? So in this case something like:

    template <typename T, size_t SeriesOrder>
    struct coeffs_C1 : boost::array<T, SeriesOrder + 1>
    {
        coeffs_C1(T const& epsilon)
        {
            // the content of evaluate_coeffs_C1()
        }
    };

    template <typename T, size_t SeriesOrder>
    struct coeffs_C3x : boost::array<T, (SeriesOrder * (SeriesOrder - 1)) / 2)>
    {
        coeffs_C3x(T const& n)
        {
            // the content of evaluate_coeffs_C3x()
        }
    };

    // etc.

this way the caller would only have to know the `SeriesOrder`. This would be the same for all coefficients. He wouldn't be forced to know how to calculate the size of array, if he should add 1 to size or not, etc. Everything would be defined in one place.

And then `sin_cos_series()` which AFAIU always calculates the series using all elements of array would take the size from `static_size` member (derived from boost::array<>)

    template <typename CT, typename Coeffs>
    inline CT sin_cos_series(CT const& sinx, CT const& cosx, Coeffs const& coeffs)
    {
        // and then:
        size_t const size = Coeffs::static_size;
        CT el0 = coeffs[0];
    }

Does it have sense or is there something about coeffiients that I don't understand?

Reply:
@awulkiew Thanks. Yes, this makes perfect sense. This will also make the code in `karney_direct.hpp` simpler. Also, `series_expansion.hpp` seems like a better place to include size computation logic for coefficient containers.